packages/next/src/server/lib/router-utils/typegen.ts

```diff
@@ -0,0 +1,194 @@
+  const allRoutes: Record<string, any> = {
+    ...routesManifest.appRoutes,
+    ...routesManifest.pageRoutes,
+    ...routesManifest.layoutRoutes,
+    // Redirect / rewrite routes are optional
+    ...((routesManifest as any).redirectRoutes ?? {}),
+    ...((routesManifest as any).rewriteRoutes ?? {}),
+  }
+
+  let paramTypes = 'type ParamMap = {\n'
+
+  // Sort routes deterministically for consistent output
+  const sortedRoutes = Object.entries(allRoutes).sort(([a], [b]) =>
+    a.localeCompare(b)
+  )
+
+  for (const [route, routeInfo] of sortedRoutes) {
+    const { groups } = routeInfo as any
+
+    // For static routes (no dynamic segments), we can produce an empty parameter map.
+    if (!isDynamicRoute(route) || Object.keys(groups ?? {}).length === 0) {
+      paramTypes += `  ${JSON.stringify(route)}: {}\n`
+      continue
+    }
+
+    let paramType = '{'
+
+    // Process each group based on its properties
+    for (const [key, group] of Object.entries(groups as Record<string, any>)) {
+      const escapedKey = JSON.stringify(key)
+      if ((group as any).repeat) {
+        // Catch-all parameters
+        if ((group as any).optional) {
+          paramType += ` ${escapedKey}?: string[];`
+        } else {
+          paramType += ` ${escapedKey}: string[];`
+        }
+      } else {
+        // Regular parameters
+        if ((group as any).optional) {
+          paramType += ` ${escapedKey}?: string;`
+        } else {
+          paramType += ` ${escapedKey}: string;`
+        }
+      }
+    }
+
+    paramType += ' }'
+
+    paramTypes += `  ${JSON.stringify(route)}: ${paramType}\n`
+  }
+
+  paramTypes += '}\n'
+  return paramTypes
+}
+
+function generateLayoutSlotMap(routesManifest: RouteTypesManifest): string {
+  let slotMap = 'type LayoutSlotMap = {\n'
+
+  // Sort routes deterministically for consistent output
+  const sortedLayoutRoutes = Object.entries(routesManifest.layoutRoutes).sort(
+    ([a], [b]) => a.localeCompare(b)
+  )
+
+  for (const [route, routeInfo] of sortedLayoutRoutes) {
+    if ('slots' in routeInfo) {
+      const slots = routeInfo.slots.sort()
+      if (slots.length > 0) {
+        slotMap += `  ${JSON.stringify(route)}: ${slots.map((slot) => JSON.stringify(slot)).join(' | ')}\n`
+      } else {
+        slotMap += `  ${JSON.stringify(route)}: never\n`
+      }
+    } else {
+      slotMap += `  ${JSON.stringify(route)}: never\n`
+    }
+  }
+
+  slotMap += '}\n'
+  return slotMap
+}
+
+export function generateRouteTypesFile(
+  routesManifest: RouteTypesManifest
+): string {
+  const routeTypes = generateRouteTypes(routesManifest)
+  const paramTypes = generateParamTypes(routesManifest)
+  const layoutSlotMap = generateLayoutSlotMap(routesManifest)
+
+  return `// This file is generated automatically by Next.js
+// Do not edit this file manually
+
+${routeTypes}
+
+${paramTypes}
+
+export type ParamsOf<P extends Routes> = ParamMap[P]
```

Comment:
Let's write out these variable names. The single-letters were originally motivated by fully generic things (e.g. `map<T>(items: T[])` or `map<T>(items: T[], mapper:(item: T) => U)`) but then everybody assumed all type parameters should have single letters hurting readability for things that aren't fully generic.

```suggestion
export type ParamsOf<Route extends Routes> = ParamMap[Route]
```

We probably don't need that indirection anyway but the general review applies: Let's make type params readable.

Reply:
Thanks for the feedback -- fixed!